crates/counter-agg/src/stable.rs

```diff
@@ -0,0 +1,62 @@
+//! These data structures are stable, meaning they may not change even in
+//! their layout in memory, as the raw bytes-in-memory are serialized and
+//! exchanged by PostgreSQL.
```

Comment:
@JLockerman can correct me if I have this wrong, but this is not actually the case for these objects.  The `MetricSummary` is only used in the internal state, which is never persisted across version upgrades.  The objects that aren't allowed to change are the `pg_type` declared final objects, `CounterSummary` and `FlatSummary` in this case.

Reply:
Oh, that would be a relief! If that's so, why do we test it? If we don't actually require stability of that serialization, we're just paying a tax every time we change it:

https://github.com/timescale/timescaledb-toolkit/blob/main/extension/src/counter_agg.rs#L972

Reply:
TLDR: you're both partially right, the situation is confusing, and the problem goes away once we stop caring about timescaledb versions older than 2.7.0

Ah this is the confusing bit: postgres has two different methods of serialziation

1. `varlena` types must exist as a flat memory buffer, and are written to disk in the same format as they exist in memory.
2. If you want to use an aggregate in parallel execution, you need to be able to serialize the transition state into some (machine-agnostic?) wire format for IPC. Notably, we used to store the output of this in continuous aggregates, though that's slated to change in timescaledb 2.7.0

@WireBaron is saying that these types aren't the first kind; they only exist in transition functions–ephemerally in memory–and therefore they can change between version. The test is checking that we can serialize this to disk stably in continuous aggregates, which is a _slightly_ weaker invariant; we can change the in-memory representation freely as long as we can keep deserializing old partials. Hopefully we'll be able to get everyone onto the new continuous aggregates format soon, and we'll be able to stop caring about the second form of stability.

Reply:
Notably this means that the following comment
```rust
//! Note that [MetricSummary] is already in violation, as it does not lock in
//! a memory representation and the Rust project makes no guarantees to
//! preserve this across releases of the compiler.  We should bump its
//! serialization version and repr(C) the new one.
```
is incorrect: the rust-memory layout doesn't really matter here, what matters is that serde guarantees layout-stability for the serialized form of a given struct definition.

Reply:
(the incorrect leap i was making was from "we directly derive the serialization form from the in-memory layout" (true) to "we serialize the actual in-memory layout" (false); the latter necessarily leads to "we need repr(C)", it's just not so! :)

Josh points out on a call that what I'm doing here is backwards: adding a second struct from which to derive (de)serialize code from when instead I should stop deriving and implement them directly. Which I'll do when I return to this, likely next week.

Thanks!